### PR TITLE
CommandMenu: inline ghost-text completion with a specified keyboard model (#970)

### LINKED ISSUE #970: CommandMenu: add semi-transparent inline autocomplete and specify the full keyboard model
Feature request, not a defect. As the user moves through options, the remainder of the highlighted option's text should appear inline after the typed characters at reduced opacity, with an explicit key accepting it.

## Current state

CommandMenu filters a listbox below the caret with no inline ghost text; trigger detection lives in `command-menu-trigger.ts` (re-exported as `detectTrigger`, `command-menu.svelte:21`); active-item state is the `caretIndex` prop (`command-menu.types.ts:30`) — separately flagged as poorly named and questionably **required** (every consumer must supply it, see #922), and it is the same state a ghost-text feature must read, so **design the two together rather than sequentially**.

## Questions that must be answered before implementation, not during

**(a) Where does the ghost text live** — a textarea/input cannot render styled inline text, so this needs either an overlay element aligned to the caret or a contenteditable surface; this is the crux and should be settled first.

**(b) Which key accepts** — Tab, Right Arrow, or Enter — and how that interacts with the listbox's own Enter-to-select.

**(c) Accessibility** — the completion must not be announced as user-entered content; it needs `aria-hidden` with real state conveyed via `aria-activedescendant` on the input plus the existing listbox semantics.

**(d) The full keyboard matrix specified up front** — Up/Down, Home/End, Escape, wrap-around, and behavior when the filtered list is empty (repo guidance is explicit about enumerating the edge-case matrix before writing a state machine).

## Notes

Autocomplete already solves adjacent problems and should be consulted for shared behavior rather than growing a second divergent implementation — see #919.

Escape currently doesn't stay dismissed (#930), which interacts with this work.

`caretIndex`'s naming and required-ness is tracked under the prop/API vocabulary epic (#922) — resolve that alongside this feature rather than locking in the current shape.

## Related

#919, #922, #930

### PR DESCRIPTION
**Pull request by stevekinney**
Closes #970

## Summary

Adds semi-transparent inline "ghost text" autocomplete to CommandMenu, gated entirely behind a new opt-in `onComplete` prop, with the full keyboard model the issue required specified up front (not discovered through review round-trips) and recorded in `command-menu.a11y.md`.

`caretIndex` is also now optional: ghost text's own caret-position need (`caretAtFieldEnd`) reads the anchor's live `selectionStart`/`selectionEnd` directly and never touches the `caretIndex` prop, which is the evidence that settles its required-ness question left open after #922 closed. When omitted, it derives from `anchor.selectionEnd`. Every existing consumer keeps passing it explicitly, so this is additive — zero behavior change for current callers.

## What I found and reused from `autocomplete/` (step 2)

`autocomplete.svelte` has **no inline-completion logic** — its `onComplete` fires only after a full-value replacement on Enter/click selection, so there was nothing to lift wholesale for ghost text's ArrowRight/Tab-accept model. Three things were still worth taking from it:

- **The `onComplete` prop name.** CommandMenu's ghost-text acceptance callback reuses this exact name for consistency, even though the semantics differ (Autocomplete replaces the whole value; CommandMenu hands back `{ value, query, remainder }` and the host appends `remainder`, matching CommandMenu's existing `onSelect` contract where the host always owns text replacement).
- **The IME composition guard shape** — `composing` state toggled by `oncompositionstart`/`oncompositionend`, checked before treating a keystroke as an accept/dismiss.
- **Confirmation that Home/End are already claimed for listbox navigation, not caret movement**, in a text-field-anchored list component. Autocomplete's `moveToBoundary` already does this; CommandMenu's shared `CommandListState.handleKeydown` does too (pre-existing, tested at `command-menu.test.ts:214,324`). This directly overrides the issue's own "Home/End caret movement (End accepts)" suggestion — see decision (b)/(d) below.

## The four recorded decisions (`command-menu.a11y.md`)

**(a) Rendering surface — `aria-hidden` overlay, not contenteditable.** A portaled `<span>` positioned via the *same* mirrored-`<div>` text measurement `caret-rect.svelte.ts` already uses to place the listbox (`getCaretRect`, called with `anchor.value.length`). Direct in-repo precedent, no new technique. The overlay's position is a synchronous `$derived`, not a second `createAnchoredOverlay`/floating-ui instance — floating-ui's `flip` middleware would, at a viewport edge, flip the overlay onto *already-typed* text, which is worse than doing nothing, and its async `computePosition` would lag the ghost by a frame every keystroke.

**(b) Acceptance keys and Enter precedence.**
- ArrowRight (unmodified, caret at field end) and Tab (unmodified) accept.
- **Enter always activates the listbox selection** — never accepts ghost text. Matches the existing shared list, which already claims unmodified Enter unconditionally.
- **End does *not* accept**, contradicting the issue's own suggestion — in-repo precedent wins: `CommandListState.handleKeydown` already claims unmodified `End` to jump the active item to the last option (pre-existing, tested). Overloading it for ghost-text acceptance would silently steal "jump to last" for the entire time a host has the feature enabled.
- Accepting fires `onComplete({ value, query, remainder })`, not `onSelect` — it completes the *typed text* (shell-tab-completion style), it does not select/commit the command.

**(c) Announcements.** The ghost span is `aria-hidden="true"`. Real state is `aria-activedescendant` + the existing listbox semantics (unchanged). `aria-autocomplete="both"` on the host field when `onComplete` is passed (vs `"list"` when it isn't) — standard ARIA vocabulary, not new. Demonstrated in the updated playground example.

**(d) Full keyboard matrix** — see the table below and the fuller version in `command-menu.a11y.md`, including two named, accepted limitations: ghost text only ever shows when the caret is at the field's end (an overlay can't render ahead of trailing text — no mid-paragraph ghost text in a multi-line textarea), and Escape's ghost-dismissal re-arms on the very next `ArrowDown` (active item changed) rather than surviving navigation to a different item.

| Key | Effect |
| --- | --- |
| ArrowUp / ArrowDown | Moves the active item (pre-existing). Ghost text updates for free — it's a derived read of `activeItemId`. |
| Home | Pre-existing, now documented: jumps to the first enabled option. Not caret movement. |
| End | Pre-existing, now documented: jumps to the last enabled option. Does **not** accept ghost text. |
| ArrowRight, unmodified, at field end | Accepts when ghost text is visible. Otherwise native caret move (no-op). |
| ArrowRight, caret mid-value | Native caret move — ghost text is never visible here. |
| Tab, unmodified | Accepts when visible, keeping focus on the anchor. Otherwise untouched (native focus traversal). |
| Shift+Tab | Never intercepted. |
| Enter | Always activates the listbox selection. Never accepts ghost text. |
| Escape, ghost visible | First stage: dismisses the ghost text only. Menu stays open. |
| Escape, ghost already hidden | Falls through to the existing listbox Escape-dismiss latch. |
| Wrap-around | Pre-existing ArrowUp/ArrowDown wrap is unchanged; ghost text follows the wrapped active item. |
| Empty filtered results | No active item ⇒ ghost text hidden. |
| IME composition | Hidden while composing (checked live via `compositionstart`/`compositionend`, and defensively again via `event.isComposing`/`keyCode === 229` at accept/dismiss time). |
| RTL field direction | Hidden (`isRightToLeftElement(anchor)`) — fails toward "no ghost text," not a mirrored or misleading render. |
| Caret not at field end | Hidden. |
| Backspace / any net-shrinking edit (typed or pasted) | Suppressed for that one keystroke only. Re-arms once the query grows again. |
| Paste that grows the query | Shown, same as typing. |

## Implementation notes

- State machine (`command-menu-inline-completion.svelte.ts`) owns *whether* to show ghost text and *what it says*; `command-menu.svelte` owns DOM wiring (live caret tracking, overlay positioning, keydown interception, portaling). Neither file exceeds ~420 lines.
- The ghost span's color comes from the `--cinder-text-disabled` CSS token in `command-menu.css`, not an inline style — an inline `color` would beat the cascade layer and defeat the "semi-transparent" point. Verified against a live static export of the playground: the ghost span's computed color (`oklch(52% 0.01 245)`) differs from the anchor's own text color (`oklch(20% 0.018 245)`), with no inline `color` in its `style` attribute.
- Full edge-case matrix mapped before coding: typing/selection/filtering races, IME composition, RTL, caret-not-at-end, paste, backspace (never re-complete on deletion) — all covered by tests.
- `caretIndex` optional-with-derivation, `onComplete` detail carries `remainder` (not just `value`) so hosts can't accidentally normalize the user's typed casing by replacing the whole token.

## Tests

- `command-menu.test.ts`: 20+ new cases covering the full matrix above (render/hide conditions, accept via ArrowRight/Tab, Enter-always-wins, End-does-not-accept, two-stage Escape, IME, RTL, backspace/paste races, disabled-feature regression, `caretIndex` omitted).
- `command-menu-inline-completion.test.ts`: pure-function coverage for `computeGhostRemainder` (case-insensitive matching, casing preservation, empty/no-match/exact-match) and `computeGhostOverlayFontStyle` (metrics only, explicitly *not* color).
- Full `command-menu` suite green (62/62). `bun run typecheck` (svelte-check) clean. oxlint + prettier + stylelint clean on all touched files. `components:check` and `exports:check` pass.

## Docs / examples

- `command-menu.a11y.md` carries the full decision record.
- `README.md`/`command-menu.schema.json` regenerated via `components:generate`.
- `slash-in-textarea.example.svelte` (playground) now wires `onComplete` and `aria-autocomplete="both"`; `command-menu.examples.json` regenerated from it, not hand-edited.

## Changeset

`minor` on `@lostgradient/cinder` — new feature surface (pre-1.0, so a prop-shape change like `caretIndex` becoming optional also ships as `minor` per this repo's convention, not `major`).

### CODE REVIEW

File: packages/components/src/components/command-menu/index.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import CommandMenu from './command-menu.svelte';
 export default CommandMenu;
 export { detectTrigger } from './command-menu-trigger.ts';
 export type {
+  CommandMenuCompletion,
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/components/command-menu/index.ts:7]: Fixed in 32e824e43: `CommandMenuCompletion` is now re-exported from `packages/components/src/index.ts` alongside the other `CommandMenu` types, keeping both public import shapes in sync. Re-ran `exports:generate`/`exports:check` — no drift, since this is a root-barrel type export and doesn't touch the generated subpath exports map.

File: packages/components/src/components/command-menu/command-menu.svelte [hunk 2]
```diff
@@ -232,17 +373,35 @@
   $effect(() => {
     if (!open || !anchor) return;
     const stopKeydown = on(anchor, 'keydown', handleKeydown);
+    const stopCompositionStart = on(anchor, 'compositionstart', () => {
+      composing = true;
+    });
+    const stopCompositionEnd = on(anchor, 'compositionend', () => {
+      composing = false;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/components/command-menu/command-menu.svelte:386]: Fixed in 32e824e43: the keydown-wiring effect's cleanup now resets `composing = false` alongside removing the composition listeners. Previously, a compositionstart with no matching compositionend (menu closed by an outside pointerdown or host-controlled dismiss while an IME composition was in flight) latched `composing` true forever, since the only thing that ever cleared it was a compositionend on listeners that had just been torn down — permanently suppressing ghost text on every later open. Added a regression test (`a composition left in flight when the menu closes does not permanently suppress ghost text`) that opens, starts composing, dismisses via outside pointerdown, reopens, and asserts ghost text renders again.

File: packages/components/src/components/command-menu/command-menu.svelte [hunk 3]
```diff
@@ -99,6 +120,67 @@
...
+      currentAnchor.selectionStart === currentAnchor.selectionEnd &&
+      currentAnchor.selectionEnd === currentAnchor.value.length
+    );
+  });
+
+  const activeValue = $derived.by(() => {
+    const id = open ? commandList.activeItemId : null;
+    if (id === null) return null;
+    const record = commandList.registrations.find((registration) => registration.id === id);
+    return record?.getValue() ?? null;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/components/command-menu/command-menu.svelte:145]: Declining this one — I think it's a false positive given how `computeGhostRemainder` and `registrationsReady` are actually specced.

For a stale registration to render *wrong* ghost text, its value would have to still pass `activeValue.toLowerCase().startsWith(query.toLowerCase())` against the *new* query while the host has already dropped it from the visible list. But if the value still prefix-matches the new query, that's a real, currently-registered item's value — the worst case is ghost text for an item the host is about to stop rendering, visible for at most one reactive flush. That's not incorrect output; it's the documented division of labor in the module header: the host owns filtering (here, `label.includes(query)`, a *different* predicate than the value-prefix check), and ghost text owns its own `startsWith` gate on `value`. A record that's actually been filtered out because it no longer matches the query can't pass that gate — the prefix check is exactly why `command-menu-inline-completion.svelte.ts`'s header comment calls this race-proof.

I also don't think gating on `registrationsReady` is available as a low-cost belt-and-suspenders here: `command-menu.svelte:281-284` calls `commandList.refreshRegistrationsReady()` on *every* query change, unconditionally, not just on ones where the registration set actually shifts. Wiring ghost text to that flag would blank it for a microtask on every keystroke — a guaranteed, visible regression traded for a hypothetical one-frame stale value that the existing prefix check already forecloses.

Leaving this thread open in case there's a concrete repro I'm missing, but not changing the implementation for it.
